cmake/Util.cmake

```diff
@@ -0,0 +1,58 @@
+# Defines a target that depends on ${FILES} and the files found by globbing
+# when using ${GLOB_PAT} and ${GLOB_DIRS}. The target will only run if any of
+# the files it depends on is changed.
+#
+# REQUIRED  - Abort if COMMAND doesn't exist.
+# GLOB_PAT  - Glob pattern to use. Only used if ${GLOB_DIRS} is specified.
+# GLOB_DIRS - The directories to recursively search for files with extension
+#             ${GLOB_PAT}.
+function(add_smart_target)
+  cmake_parse_arguments(SMART
+    "REQUIRED"
+    "TARGET;COMMAND;GLOB_PAT"
+    "FLAGS;FILES;GLOB_DIRS"
+    ${ARGN}
+  )
+
+  string(TOUPPER ${SMART_COMMAND} COMMAND_UPPER)
+  set(PRG_VAR ${COMMAND_UPPER}_PRG)
+  find_program(${PRG_VAR} ${SMART_COMMAND})
+  set(PRG ${${PRG_VAR}})
+
+  if(NOT PRG)
+    add_custom_target(${SMART_TARGET})
+    if(SMART_REQUIRED)
+      add_custom_command(TARGET ${SMART_TARGET}
+        COMMAND ${CMAKE_COMMAND} -E echo "${SMART_TARGET}: ${SMART_COMMAND} not found"
+        COMMAND false)
+    else()
+      add_custom_command(TARGET ${SMART_TARGET}
+        COMMAND ${CMAKE_COMMAND} -E echo "${SMART_TARGET} SKIP: ${SMART_COMMAND} not found")
+    endif()
+
+    return()
+  endif()
+
+  foreach(gd ${SMART_GLOB_DIRS})
+    file(GLOB_RECURSE globfiles ${PROJECT_SOURCE_DIR}/${gd}/${SMART_GLOB_PAT})
+    list(APPEND SMART_FILES ${globfiles})
+  endforeach()
+
+  set(touch_dir ${TOUCHES_DIR}/${SMART_TARGET})
+  file(MAKE_DIRECTORY ${touch_dir})
+
+  foreach(f ${SMART_FILES})
```

Comment:
I see that this is lifting the pattern already used for the old "lintc" target: https://github.com/neovim/neovim/blob/12c62ddea64159f812682d6cccdb3bf0ff50d035/src/nvim/CMakeLists.txt#L788-L804

That's good. There might be a cmake feature that does this, but this new function looks pretty low-risk.

Later we could try to convert the "lintc" target.